daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala

```diff
@@ -903,7 +903,7 @@ sealed trait DIElement
    */
   final def maybeIsNilled: MaybeBoolean = {
     if (!_isNilledSet) MaybeBoolean.Nope
-    MaybeBoolean(_isNilled)
+    else MaybeBoolean(_isNilled)
```

Comment:
This looks to be a bug fix. We should consider adding a test that would have caught it, since there obviously wasn't one before.

Reply:
Yeah, this definitely feels like a bug, but code coverage does show both lines as being hit. I'll see if I can figure out when  this is used and see if there's a way to reproduce whatever bug already existed.

Reply:
After some investigation, while this code is logically wrong, we always use ``maybeIsNilled`` in a manner that doesn't cause any bugs. Previously, this function could only return two values: ``MaybeBoolean.True`` or ``MaybeBoolean.false``. If nilled hadn't been set yet, it would return ``MaybeBoolean.False`` since false is the default value of _isNilled.

Whenever ``maybeIsNilled`` is used, we always have something to the affect of:
```
if (maybeIsNilled.isDefined && maybeIsNilled.get == true)
```
The logic is the same regardless if we return MaybeBoolean.Nope or MaybeBoolean.False. The only place this could cause a bug is we actually cared if maybeIsNilled hadn't been set yet, or or it had be explicitly set to false. We don't ever do that, so there isn't currently a bug.

That said, I'm wondering if we should even have maybeIsNilled. The above conditional seems wrong to me. If maybeIsNilled hasn't been set yet, we shouldn't be assuming it's false. In fact, I can't imagine a case where want to know if something is nillled but don't have an answer yet. Seems like this should be able to go away and we should just always be able to ask isNilled and get an answer.

Reply:
So this distinction of MaybeBoolean.Nope vs. true/false I can see it being needed for implementing unparser default values, or possibly even parser default values. 

It may have been put in, in anticipation of those features.

Also the expression language currently has no way to provide a "nil" value for a nillable element. This is likely a shortcoming. 

Were that to be added, we would need exactly this mechanism to detect completion of unparser expression evaluation.

I think we should leave this functionality in place for now. Albeit, it was broken.

Reply:
Sounds reasonable to me.